### PR TITLE
Changes for merging shine/icfp20 branch

### PR DESCRIPTION
**Pull request by michel-steuwer**

### CODE REVIEW

**Review by bastian-koepcke — DISMISSED** [review]

File: src/main/scala/elevate/rise/rules/lowering.scala [hunk 1]
```diff
@@ -245,7 +248,9 @@ object lowering {
 
   // todo gotta use a normalform for introducing copies! e.g., if we have two reduce primitives
   def lowerToC(implicit ev: Traversable[Rise]): Strategy[Rise] =
-    addRequiredCopies `;` `try`(bottomUp(copyAfterReduceInit)) `;` specializeSeq
+    debug[Rise]("before lowering") `;`
+    addRequiredCopies `;` specializeSeq `;` debug("after lowering")
```
Comment (Bastacyclop) [inline src/main/scala/elevate/rise/rules/lowering.scala]: are we keeping these debug prints in the PR?

File: src/main/scala/elevate/rise/rules/lowering.scala [hunk 2]
```diff
@@ -259,7 +264,7 @@ object lowering {
 
     def apply(e: Rise): RewriteResult[Rise] = e match {
       case reduceResult@App(App(App(ReduceX(), _), _), _) =>
```
Comment (Bastacyclop) [inline src/main/scala/elevate/rise/rules/lowering.scala:264]: is the `reduceResult` name needed at all?

File: src/main/scala/elevate/rise/rules/movement.scala [hunk 3]
```diff
@@ -96,6 +96,18 @@ object movement {
...
+//  case object `*f >> S -> S >> **f` extends Strategy[Rise] {
+//    def apply(e: Rise): RewriteResult[Rise] = e match {
+//      case App(
+//      s,
+//      App(App(Map(), f), y)) if isSplitOrSlide(s) =>
+//        Success((typed(y) |> untyped(s) |> map(map(f))) :: e.t)
+//      case _ => Failure(mapFBeforeSlide)
+//    }
+//    override def toString = "mapFBeforeSlide"
+//  }
```
Comment (Bastacyclop) [inline src/main/scala/elevate/rise/rules/movement.scala]: should we keep this commented out code?

**Review by Bastacyclop — DISMISSED** [review]

File: src/main/scala/elevate/rise/rules/lowering.scala (hunk 1 above)
Comment (michel-steuwer) [reply src/main/scala/elevate/rise/rules/lowering.scala]: I will removed them

File: src/main/scala/elevate/rise/rules/lowering.scala (hunk 2 above)
Comment (michel-steuwer) [reply src/main/scala/elevate/rise/rules/lowering.scala:264]: The name is used here to be able to refer to the type of the entire pattern (i.e. `reduceResult.t`)

File: src/main/scala/elevate/rise/rules/movement.scala (hunk 3 above)
Comment (michel-steuwer) [reply src/main/scala/elevate/rise/rules/movement.scala]: I remove this

File: src/main/scala/elevate/rise/rules/lowering.scala (hunk 2 above)
Comment (Bastacyclop) [reply src/main/scala/elevate/rise/rules/lowering.scala:264]: I mean we can use `e`, correct?

**Review by Bastacyclop — APPROVED** [review]

**Pull request merged**